File: src/core/dedup.ts
async function sha256Hex(value: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(value));
  return [...new Uint8Array(digest)].map((n) => n.toString(16).padStart(2, '0')).join('');
}

export function normalizeText(value?: string | null): string {
  if (!value) return '';
  return value.trim().toLowerCase().split(/\s+/).join(' ');
}

export async function buildContentHash(title: string | null | undefined, url: string | null | undefined, source: string): Promise<string> {
  const basis = [normalizeText(source), normalizeText(title), normalizeText(url)].join('|');
  return sha256Hex(basis);
}

File: ingestion/dedup.py
from __future__ import annotations

import hashlib


def normalize_text(value: str | None) -> str:
    if not value:
        return ""
    return " ".join(value.strip().lower().split())


def build_content_hash(title: str | None, url: str | None, source: str) -> str:
    """Hash title + url + source to catch near-duplicate API inconsistencies."""

    basis = "|".join(
        [
            normalize_text(source),
            normalize_text(title),
            normalize_text(url),
        ]
    )
    return hashlib.sha256(basis.encode("utf-8")).hexdigest()
